### PR TITLE
Fix race in tsdb

### PR DESCRIPTION
**Pull request by gaurav1086**
When a go loop launches goroutines, it is a common mistake to assume
that the routines can reference the loop value from their specific loop
iteration. Actually, all goroutines share the same reference to the loop
veriable, and they typically all see the last value. To correct this, the loop variable must be captured as a parameter to the anonymous func.

### CODE REVIEW

File: tsdb/index/postings.go [hunk 1]
```diff
@@ -168,12 +168,12 @@ func (p *MemPostings) EnsureOrder() {
 	wg.Add(n)
 
 	for i := 0; i < n; i++ {
-		go func() {
+		go func(i interface{}) {
```
Comment (bwplotka) [inline tsdb/index/postings.go:171]:
```suggestion
		go func(i int) {
```

File: tsdb/index/postings.go [hunk 2]
```diff
@@ -168,12 +168,12 @@ func (p *MemPostings) EnsureOrder() {
...
 			for l := range workc {
 				sort.Slice(l, func(i, j int) bool { return l[i] < l[j] })
```
Comment (roidelapluie) [inline tsdb/index/postings.go:173]:
```suggestion
				sort.Slice(l, func(a, b int) bool { return l[a] < l[b] })
```

File: tsdb/index/postings.go (hunk 1 above)
Comment (roidelapluie) [inline tsdb/index/postings.go:171]:
```suggestion
		go func() {
```
That will need to be reverted

File: tsdb/index/postings.go (hunk 2 above)
Comment (bwplotka) [reply tsdb/index/postings.go]: or this

File: discovery/kubernetes/client_metrics.go [hunk 3]
```diff
@@ -160,24 +159,3 @@ func (clientGoWorkqueueMetricsProvider) NewRetriesMetric(name string) workqueue.
 	// Retries are not used so the metric is omitted.
 	return noopMetric{}
 }
-func (clientGoWorkqueueMetricsProvider) NewDeprecatedDepthMetric(name string) workqueue.GaugeMetric {
```
Comment (bwplotka) [inline discovery/kubernetes/client_metrics.go:163]: unrelated change?

File: tsdb/head.go [hunk 4]
```diff
@@ -681,9 +681,11 @@ func (h *Head) Truncate(mint int64) (err error) {
...
 		return nil // no segments yet.
 	}
-	// The lower third of segments should contain mostly obsolete samples.
-	// If we have less than three segments, it's not worth checkpointing yet.
-	last = first + (last-first)/3
+	// The lower two thirds of segments should contain mostly obsolete samples.
+	// If we have less than two segments, it's not worth checkpointing yet.
+	// With the default 2h blocks, this will keeping up to around 3h worth
+	// of WAL segments.
+	last = first + (last-first)*2/3
```
Comment (codesome) [inline tsdb/head.go:688]: Could you rebase your PR? This change is already in master.

**Comment by gaurav1086** [comment]
Will create a fresh PR. Closing this one.
